src/auto_note.cpp

```diff
@@ -0,0 +1,454 @@
+#include "output.h"
+#include "debug.h"
+#include "filesystem.h"
+#include "translations.h"
+#include "avatar.h"
+#include "map_extras.h"
+
+namespace auto_notes
+{
+std::string auto_note_settings::build_save_path() const
+{
+    return g->get_player_base_save_path() + ".ano.json";
+}
+
+void auto_note_settings::clear()
+{
+    this->autoNoteEnabled.clear();
+}
+
+bool auto_note_settings::save()
+{
+    // Check if the player is already saved.
+    if( !file_exist( g->get_player_base_save_path() + ".sav" ) ) {
+        return true;    // Nothing to do here
+    }
+
+    // Serialize contents of map ext set as array of strings
+    return write_to_file( this->build_save_path(), [&]( std::ostream & fstr ) {
+        JsonOut jout{ fstr, true };
+
+        jout.start_object();
+
+        jout.member( "enabled" );
+
+        jout.start_array();
+        for( const std::string &entry : this->autoNoteEnabled ) {
+            jout.write( entry );
+        }
+        jout.end_array();
+
+        jout.member( "discovered" );
+
+        jout.start_array();
+        for( const std::string &entry : this->discovered ) {
+            jout.write( entry );
+        }
+        jout.end_array();
+
+        jout.end_object();
+
+    }, _( "auto notes configuration" ) );
+}
+
+void auto_note_settings::load()
+{
+    // Clear all remaining settings
+    this->clear();
+
+    // Lambda used inside the read from file call
+    const auto parseJson = [&]( JsonIn & jin ) {
+        jin.start_object();
+
+        while( !jin.end_object() ) {
+            const std::string name = jin.get_member_name();
+
+            if( name == "enabled" ) {
+                jin.start_array();
+                while( !jin.end_array() ) {
+                    const std::string entry = jin.get_string();
+                    this->autoNoteEnabled.insert( entry );
+                }
+            } else if( name == "discovered" ) {
+                jin.start_array();
+                while( !jin.end_array() ) {
+                    const std::string entry = jin.get_string();
+                    this->discovered.insert( entry );
+                }
+            } else {
+                jin.skip_value();
+            }
+
+        }
+    };
+
+    // Actually read save file
+    if( !read_from_file_optional_json( this->build_save_path(), parseJson ) ) {
+        // If loading from JSON failed, fall back to default initialization
+        this->default_initialize();
+        this->save();
+    }
+}
+
+void auto_note_settings::default_initialize()
+{
+    // Perform clear for good measure
+    this->clear();
+
+    // Fetch all know map extra types and iterate over them
+    for( auto extra : MapExtras::mapExtraFactory().get_all() ) {
+        if( extra.autonote ) {
+            this->autoNoteEnabled.insert( extra.id.str() );
+        }
+    }
+}
+
+void auto_note_settings::set_discovered( const string_id<map_extra> &mapExtId )
+{
+    this->discovered.insert( mapExtId.str() );
```

Comment:
`discovered` should contain objects of type `string_id<map_extra>`, so you don't have to convert between a raw string an a `string_id<map_extra>` that often and at arbitrary places.